Implement Key for Pubkey

https://github.com/project-serum/anchor/pull/1127/files#diff-c30312f654589dc8120c3ca5d75db6816b7825627c10c7c4b485215991bd54e7L237 removed the implementation.

When adding the trait implementation, we get the following error

```
error[E0119]: conflicting implementations of trait `Key` for type `solana_program::pubkey::Pubkey`
   --> lang/src/lib.rs:232:1
    |
226 |   impl Key for Pubkey {
    |   ------------------- first implementation here
...
232 | / impl<'info, T> Key for T
233 | | where
234 | |     T: AsRef<AccountInfo<'info>>,
235 | | {
...   |
238 | |     }
239 | | }
    | |_^ conflicting implementation for `solana_program::pubkey::Pubkey`
  ```

@skrrb sorry, I needed to mention that error happens in macros:

```rust
    #[account(mut, associated_token::mint = MINT_PUBKEY, associated_token::authority = account.signer)]
```

```rust
error[E0599]: the method `key` exists for struct `anchor_lang::prelude::Pubkey`, but its trait bounds were not satisfied
   --> programs/my-stupid-program/src/lib.rs:132:10
    |
132 | #[derive(Accounts)]
    |          ^^^^^^^^ method cannot be called on `anchor_lang::prelude::Pubkey` due to unsatisfied trait bounds
    |
   ::: /home/kirill/.cargo/registry/src/github.com-1ecc6299db9ec823/solana-program-1.8.5/src/pubkey.rs:70:1
    |
70  | pub struct Pubkey([u8; 32]);
    | ----------------------------
    | |
    | doesn't satisfy `_: std::convert::AsRef<anchor_lang::prelude::AccountInfo<'_>>`
    | doesn't satisfy `anchor_lang::prelude::Pubkey: anchor_lang::Key`
    |
    = note: the following trait bounds were not satisfied:
            `anchor_lang::prelude::Pubkey: std::convert::AsRef<anchor_lang::prelude::AccountInfo<'_>>`
            which is required by `anchor_lang::prelude::Pubkey: anchor_lang::Key`
    = note: this error originates in the derive macro `Accounts` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@ incremented for features.
 
 * ts: Fix the loss of strict typing using the `methods` namespace on builder functions ([#1539](https://github.com/project-serum/anchor/pull/1539)).
 * spl: Update `spl/governance` to use new errors ([#1582](https://github.com/project-serum/anchor/pull/1582)).
+* lang: implement `Key` for `Pubkey` again, so `associated_token::*` constraints can use pubkey targets again ([#1601](https://github.com/project-serum/anchor/pull/1601))
 
 ### Breaking
 

--- a/lang/src/accounts/account.rs
+++ b/lang/src/accounts/account.rs
@@ -1,10 +1,14 @@
 //! Account container that checks ownership on deserialization.
 
 use crate::error::ErrorCode;
-use crate::*;
+use crate::{
+    AccountDeserialize, AccountSerialize, Accounts, AccountsClose, AccountsExit, Key, Owner,
+    Result, ToAccountInfos, ToAccountMetas, ToAccountInfo
+};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;
+use solana_program::system_program;
 use std::collections::BTreeMap;
 use std::fmt;
 use std::ops::{Deref, DerefMut};
@@ -234,7 +238,7 @@ impl<'info, T: AccountSerialize + AccountDeserialize + Owner + Clone + fmt::Debu
     }
 }
 
-impl<'a, T: AccountSerialize + AccountDeserialize + Owner + Clone> Account<'a, T> {
+impl<'a, T: AccountSerialize + AccountDeserialize + crate::Owner + Clone> Account<'a, T> {
     fn new(info: AccountInfo<'a>, account: T) -> Account<'a, T> {
         Self { info, account }
     }
@@ -408,5 +412,11 @@ impl<'a, T: AccountSerialize + AccountDeserialize + Owner + Clone> DerefMut for 
             panic!();
         }
         &mut self.account
+    }
+}
+
+impl<'info, T: AccountSerialize + AccountDeserialize + Owner + Clone> Key for Account<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.info.key
     }
 }

--- a/lang/src/accounts/account.rs
+++ b/lang/src/accounts/account.rs
@@ -3,7 +3,7 @@
 use crate::error::ErrorCode;
 use crate::{
     AccountDeserialize, AccountSerialize, Accounts, AccountsClose, AccountsExit, Key, Owner,
-    Result, ToAccountInfos, ToAccountMetas, ToAccountInfo
+    Result, ToAccountInfo, ToAccountInfos, ToAccountMetas,
 };
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;

--- a/lang/src/accounts/account_info.rs
+++ b/lang/src/accounts/account_info.rs
@@ -3,7 +3,7 @@
 //! should be used instead.
 
 use crate::error::ErrorCode;
-use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas};
+use crate::{Accounts, AccountsExit, Key, Result, ToAccountInfos, ToAccountMetas};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;
@@ -43,3 +43,9 @@ impl<'info> ToAccountInfos<'info> for AccountInfo<'info> {
 }
 
 impl<'info> AccountsExit<'info> for AccountInfo<'info> {}
+
+impl<'info> Key for AccountInfo<'info> {
+    fn key(&self) -> Pubkey {
+        *self.key
+    }
+}

--- a/lang/src/accounts/account_loader.rs
+++ b/lang/src/accounts/account_loader.rs
@@ -2,7 +2,7 @@
 
 use crate::error::ErrorCode;
 use crate::{
-    Accounts, AccountsClose, AccountsExit, Owner, Result, ToAccountInfo, ToAccountInfos,
+    Accounts, AccountsClose, AccountsExit, Key, Owner, Result, ToAccountInfo, ToAccountInfos,
     ToAccountMetas, ZeroCopy,
 };
 use arrayref::array_ref;
@@ -264,5 +264,11 @@ impl<'info, T: ZeroCopy + Owner> AsRef<AccountInfo<'info>> for AccountLoader<'in
 impl<'info, T: ZeroCopy + Owner> ToAccountInfos<'info> for AccountLoader<'info, T> {
     fn to_account_infos(&self) -> Vec<AccountInfo<'info>> {
         vec![self.acc_info.clone()]
+    }
+}
+
+impl<'info, T: ZeroCopy + Owner> Key for AccountLoader<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.acc_info.key
     }
 }

--- a/lang/src/accounts/cpi_account.rs
+++ b/lang/src/accounts/cpi_account.rs
@@ -119,3 +119,10 @@ where
         Self::new(a.to_account_info(), Box::new(a.into_inner()))
     }
 }
+
+#[allow(deprecated)]
+impl<'info, T: AccountDeserialize + Clone> Key for CpiAccount<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.info.key
+    }
+}

--- a/lang/src/accounts/cpi_state.rs
+++ b/lang/src/accounts/cpi_state.rs
@@ -2,7 +2,7 @@ use crate::error::ErrorCode;
 #[allow(deprecated)]
 use crate::{accounts::state::ProgramState, context::CpiStateContext};
 use crate::{
-    AccountDeserialize, AccountSerialize, Accounts, AccountsExit, Result, ToAccountInfos,
+    AccountDeserialize, AccountSerialize, Accounts, AccountsExit, Key, Result, ToAccountInfos,
     ToAccountMetas,
 };
 use solana_program::account_info::AccountInfo;
@@ -138,4 +138,11 @@ impl<'info, T: AccountSerialize + AccountDeserialize + Clone> DerefMut for CpiSt
 impl<'info, T: AccountSerialize + AccountDeserialize + Clone> AccountsExit<'info>
     for CpiState<'info, T>
 {
+}
+
+#[allow(deprecated)]
+impl<'info, T: AccountSerialize + AccountDeserialize + Clone> Key for CpiState<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.inner.info.key
+    }
 }

--- a/lang/src/accounts/loader.rs
+++ b/lang/src/accounts/loader.rs
@@ -1,7 +1,7 @@
 use crate::error::ErrorCode;
 use crate::{
-    Accounts, AccountsClose, AccountsExit, Result, ToAccountInfo, ToAccountInfos, ToAccountMetas,
-    ZeroCopy,
+    Accounts, AccountsClose, AccountsExit, Key, Result, ToAccountInfo, ToAccountInfos,
+    ToAccountMetas, ZeroCopy,
 };
 use arrayref::array_ref;
 use solana_program::account_info::AccountInfo;
@@ -211,5 +211,12 @@ impl<'info, T: ZeroCopy> AsRef<AccountInfo<'info>> for Loader<'info, T> {
 impl<'info, T: ZeroCopy> ToAccountInfos<'info> for Loader<'info, T> {
     fn to_account_infos(&self) -> Vec<AccountInfo<'info>> {
         vec![self.acc_info.clone()]
+    }
+}
+
+#[allow(deprecated)]
+impl<'info, T: ZeroCopy> Key for Loader<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.acc_info.key
     }
 }

--- a/lang/src/accounts/program.rs
+++ b/lang/src/accounts/program.rs
@@ -1,7 +1,8 @@
 //! Type validating that the account is the given Program
-
 use crate::error::ErrorCode;
-use crate::*;
+use crate::{
+    AccountDeserialize, Accounts, AccountsExit, Id, Key, Result, ToAccountInfos, ToAccountMetas,
+};
 use solana_program::account_info::AccountInfo;
 use solana_program::bpf_loader_upgradeable::{self, UpgradeableLoaderState};
 use solana_program::instruction::AccountMeta;
@@ -178,3 +179,9 @@ impl<'info, T: Id + Clone> Deref for Program<'info, T> {
 }
 
 impl<'info, T: AccountDeserialize + Id + Clone> AccountsExit<'info> for Program<'info, T> {}
+
+impl<'info, T: AccountDeserialize + Id + Clone> Key for Program<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.info.key
+    }
+}

--- a/lang/src/accounts/program_account.rs
+++ b/lang/src/accounts/program_account.rs
@@ -2,7 +2,7 @@
 use crate::accounts::cpi_account::CpiAccount;
 use crate::error::ErrorCode;
 use crate::{
-    AccountDeserialize, AccountSerialize, Accounts, AccountsClose, AccountsExit, Result,
+    AccountDeserialize, AccountSerialize, Accounts, AccountsClose, AccountsExit, Key, Result,
     ToAccountInfo, ToAccountInfos, ToAccountMetas,
 };
 use solana_program::account_info::AccountInfo;
@@ -181,5 +181,12 @@ where
 {
     fn from(a: CpiAccount<'info, T>) -> Self {
         Self::new(a.to_account_info(), Deref::deref(&a).clone())
+    }
+}
+
+#[allow(deprecated)]
+impl<'info, T: AccountSerialize + AccountDeserialize + Clone> Key for ProgramAccount<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.inner.info.key
     }
 }

--- a/lang/src/accounts/signer.rs
+++ b/lang/src/accounts/signer.rs
@@ -1,6 +1,6 @@
 //! Type validating that the account signed the transaction
 use crate::error::ErrorCode;
-use crate::*;
+use crate::{Accounts, AccountsExit, Key, Result, ToAccountInfos, ToAccountMetas};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;
@@ -101,5 +101,11 @@ impl<'info> Deref for Signer<'info> {
 
     fn deref(&self) -> &Self::Target {
         &self.info
+    }
+}
+
+impl<'info> Key for Signer<'info> {
+    fn key(&self) -> Pubkey {
+        *self.info.key
     }
 }

--- a/lang/src/accounts/state.rs
+++ b/lang/src/accounts/state.rs
@@ -3,7 +3,7 @@ use crate::accounts::cpi_account::CpiAccount;
 use crate::error::ErrorCode;
 use crate::{
     AccountDeserialize, AccountSerialize, Accounts, AccountsExit, Result, ToAccountInfo,
-    ToAccountInfos, ToAccountMetas,
+    ToAccountInfos, ToAccountMetas, Key,
 };
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
@@ -159,4 +159,10 @@ pub fn address(program_id: &Pubkey) -> Pubkey {
     let seed = PROGRAM_STATE_SEED;
     let owner = program_id;
     Pubkey::create_with_seed(&base, seed, owner).unwrap()
+}
+
+impl<'info, T: AccountSerialize + AccountDeserialize + Clone> Key for ProgramState<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.inner.info.key
+    }
 }

--- a/lang/src/accounts/state.rs
+++ b/lang/src/accounts/state.rs
@@ -2,8 +2,8 @@
 use crate::accounts::cpi_account::CpiAccount;
 use crate::error::ErrorCode;
 use crate::{
-    AccountDeserialize, AccountSerialize, Accounts, AccountsExit, Result, ToAccountInfo,
-    ToAccountInfos, ToAccountMetas, Key,
+    AccountDeserialize, AccountSerialize, Accounts, AccountsExit, Key, Result, ToAccountInfo,
+    ToAccountInfos, ToAccountMetas,
 };
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;

--- a/lang/src/accounts/system_account.rs
+++ b/lang/src/accounts/system_account.rs
@@ -82,3 +82,9 @@ impl<'info> Deref for SystemAccount<'info> {
         &self.info
     }
 }
+
+impl<'info> Key for SystemAccount<'info> {
+    fn key(&self) -> Pubkey {
+        *self.info.key
+    }
+}

--- a/lang/src/accounts/sysvar.rs
+++ b/lang/src/accounts/sysvar.rs
@@ -1,7 +1,7 @@
 //! Type validating that the account is a sysvar and deserializing it
 
 use crate::error::ErrorCode;
-use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas, Key};
+use crate::{Accounts, AccountsExit, Key, Result, ToAccountInfos, ToAccountMetas};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;

--- a/lang/src/accounts/sysvar.rs
+++ b/lang/src/accounts/sysvar.rs
@@ -1,7 +1,7 @@
 //! Type validating that the account is a sysvar and deserializing it
 
 use crate::error::ErrorCode;
-use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas};
+use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas, Key};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;
@@ -114,3 +114,9 @@ impl<'a, T: solana_program::sysvar::Sysvar> DerefMut for Sysvar<'a, T> {
 }
 
 impl<'info, T: solana_program::sysvar::Sysvar> AccountsExit<'info> for Sysvar<'info, T> {}
+
+impl<'info, T: solana_program::sysvar::Sysvar> Key for Sysvar<'info, T> {
+    fn key(&self) -> Pubkey {
+        *self.info.key
+    }
+}

--- a/lang/src/accounts/unchecked_account.rs
+++ b/lang/src/accounts/unchecked_account.rs
@@ -2,7 +2,7 @@
 //! that no checks are performed
 
 use crate::error::ErrorCode;
-use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas, Key};
+use crate::{Accounts, AccountsExit, Key, Result, ToAccountInfos, ToAccountMetas};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;

--- a/lang/src/accounts/unchecked_account.rs
+++ b/lang/src/accounts/unchecked_account.rs
@@ -2,7 +2,7 @@
 //! that no checks are performed
 
 use crate::error::ErrorCode;
-use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas};
+use crate::{Accounts, AccountsExit, Result, ToAccountInfos, ToAccountMetas, Key};
 use solana_program::account_info::AccountInfo;
 use solana_program::instruction::AccountMeta;
 use solana_program::pubkey::Pubkey;
@@ -66,5 +66,11 @@ impl<'info> Deref for UncheckedAccount<'info> {
 
     fn deref(&self) -> &Self::Target {
         &self.0
+    }
+}
+
+impl<'info> Key for UncheckedAccount<'info> {
+    fn key(&self) -> Pubkey {
+        *self.0.key
     }
 }

--- a/lang/src/lib.rs
+++ b/lang/src/lib.rs
@@ -223,12 +223,9 @@ pub trait Key {
     fn key(&self) -> Pubkey;
 }
 
-impl<'info, T> Key for T
-where
-    T: AsRef<AccountInfo<'info>>,
-{
+impl Key for Pubkey {
     fn key(&self) -> Pubkey {
-        *self.as_ref().key
+        *self
     }
 }
 

--- a/lang/syn/src/codegen/accounts/constraints.rs
+++ b/lang/syn/src/codegen/accounts/constraints.rs
@@ -558,7 +558,7 @@ fn generate_constraint_seeds(f: &Field, c: &ConstraintSeedsGroup) -> proc_macro2
         .program_seed
         .clone()
         // If they specified a seeds::program to use when deriving the PDA, use it.
-        .map(|program_id| quote! { #program_id })
+        .map(|program_id| quote! { #program_id.key() })
         // Otherwise fall back to the current program's program_id.
         .unwrap_or(quote! { program_id });
 

